### PR TITLE
watch: Introduce `Watch::then_stream`

### PR DESCRIPTION
**Pull request by olix0r**
It's not unusual to want to lazily derive values from a `Watch`. For
example, when watching a resource (file, directory, etc) for updates, an
application may want to rebind services, etc.

This change introduces `Watch::then_stream` that converts a `Watch` into
a stream of values derived from borrows on the Watch. The `Then` trait is
introduced so that `ThenStream` types may be named concretely.

### CODE REVIEW

File: futures-watch/src/then_stream.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/// Maps borrowed references to `T` into an `Item`.
+pub trait Then<T> {
+    /// The output type.
+    type Output;
+
+    /// What you get when Map fails.
+    type Error;
+
+    /// Produces a new Output value.
+    fn then(&mut self, t: &T) -> Result<Self::Output, Self::Error>;
```
Comment (seanmonstar) [inline futures-watch/src/then_stream.rs]: In futures, `then` receives a `Result<T, E>`. I think if we follow that, then the `Error` enum below won't be needed either, as the `Then` impl can decide how to translate a `WatchError`.

**Comment by olix0r** [comment]
ci failure is due to busted nightly:

```
$ curl -sSf https://sh.rustup.rs | sh -s -- --default-toolchain=$TRAVIS_RUST_VERSION -y
info: downloading installer
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: latest update on 2018-06-22, rust version 1.28.0-nightly (662c70a59 2018-06-21)
info: downloading component 'rustc'
error: failed to parse url: 1.28.0-nightly (662c70a59 2018-06-21)/2018-06-22/rustc-nightly-x86_64-unknown-linux-gnu.tar.xz
```

**Review by carllerche — APPROVED** [review]
LGTM

**Comment by carllerche** [comment]
#7 will fix nightly.

**Pull request merged**